File: frontend/src/types/location.ts
export type ButwalLocation =
  | 'Traffic Chowk, Butwal'
  | 'Golpark, Butwal'
  | 'Devinagar, Butwal'
  | 'Kalikanagar, Butwal'
  | 'Milanchowk, Butwal'
  | 'Chauraha, Butwal'
  | 'Deepnagar, Butwal'
  | 'Tamnagar, Butwal'
  | 'Nayagaon, Butwal'
  | 'Manigram, Butwal'
  | 'Drivertole, Butwal'
  | 'Yogikuti, Butwal'
  | 'Belbas, Butwal'
  | 'Sukhanagar, Butwal'
  | 'Amarpath, Butwal'
  | 'Haatbazaar, Butwal'
  | 'Motipur, Butwal'
  | 'Semlar, Butwal'
  | 'Other, Butwal'

export interface LocationItem {
  id: string
  name: ButwalLocation | string
  area: string
  city: string
  province: string
}

export const BUTWAL_LOCATIONS: readonly ButwalLocation[] = [
  'Traffic Chowk, Butwal',
  'Golpark, Butwal',
  'Devinagar, Butwal',
  'Kalikanagar, Butwal',
  'Milanchowk, Butwal',
  'Chauraha, Butwal',
  'Deepnagar, Butwal',
  'Tamnagar, Butwal',
  'Nayagaon, Butwal',
  'Manigram, Butwal',
  'Drivertole, Butwal',
  'Yogikuti, Butwal',
  'Belbas, Butwal',
  'Sukhanagar, Butwal',
  'Amarpath, Butwal',
  'Haatbazaar, Butwal',
  'Motipur, Butwal',
  'Semlar, Butwal',
  'Other, Butwal',
] as const

export const SUPPORTED_LOCATIONS: readonly LocationItem[] = [
  {
    id: 'traffic-chowk',
    name: 'Traffic Chowk, Butwal',
    area: 'Traffic Chowk',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'golpark',
    name: 'Golpark, Butwal',
    area: 'Golpark',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'devinagar',
    name: 'Devinagar, Butwal',
    area: 'Devinagar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'kalikanagar',
    name: 'Kalikanagar, Butwal',
    area: 'Kalikanagar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'milanchowk',
    name: 'Milanchowk, Butwal',
    area: 'Milanchowk',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'chauraha',
    name: 'Chauraha, Butwal',
    area: 'Chauraha',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'deepnagar',
    name: 'Deepnagar, Butwal',
    area: 'Deepnagar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'tamnagar',
    name: 'Tamnagar, Butwal',
    area: 'Tamnagar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'nayagaon',
    name: 'Nayagaon, Butwal',
    area: 'Nayagaon',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'manigram',
    name: 'Manigram, Butwal',
    area: 'Manigram',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'drivertole',
    name: 'Drivertole, Butwal',
    area: 'Drivertole',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'yogikuti',
    name: 'Yogikuti, Butwal',
    area: 'Yogikuti',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'belbas',
    name: 'Belbas, Butwal',
    area: 'Belbas',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'sukhanagar',
    name: 'Sukhanagar, Butwal',
    area: 'Sukhanagar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'amarpath',
    name: 'Amarpath, Butwal',
    area: 'Amarpath',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'haatbazaar',
    name: 'Haatbazaar, Butwal',
    area: 'Haatbazaar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'motipur',
    name: 'Motipur, Butwal',
    area: 'Motipur',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'semlar',
    name: 'Semlar, Butwal',
    area: 'Semlar',
    city: 'Butwal',
    province: 'Lumbini',
  },
  {
    id: 'other-butwal',
    name: 'Other, Butwal',
    area: 'Other',
    city: 'Butwal',
    province: 'Lumbini',
  },
] as const

export type AppLocation = ButwalLocation | string

File: backend/internal/models/location.go
package models

import (
	"strings"

	"codeberg.org/amritxyz/melo/internal/utils"
)

// Supported cities and provinces in Nepal
const (
	CityButwal      = "Butwal"
	ProvinceLumbini = "Lumbini"
)

// LocationItem represents a structured location entry designed for extensible Nepal coverage.
type LocationItem struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Area     string `json:"area"`
	City     string `json:"city"`
	Province string `json:"province"`
}

// ButwalLocations contains the initial curated list of locations in Butwal.
var ButwalLocations = []LocationItem{
	{ID: "traffic-chowk", Name: "Traffic Chowk, Butwal", Area: "Traffic Chowk", City: CityButwal, Province: ProvinceLumbini},
	{ID: "golpark", Name: "Golpark, Butwal", Area: "Golpark", City: CityButwal, Province: ProvinceLumbini},
	{ID: "devinagar", Name: "Devinagar, Butwal", Area: "Devinagar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "kalikanagar", Name: "Kalikanagar, Butwal", Area: "Kalikanagar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "milanchowk", Name: "Milanchowk, Butwal", Area: "Milanchowk", City: CityButwal, Province: ProvinceLumbini},
	{ID: "chauraha", Name: "Chauraha, Butwal", Area: "Chauraha", City: CityButwal, Province: ProvinceLumbini},
	{ID: "deepnagar", Name: "Deepnagar, Butwal", Area: "Deepnagar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "tamnagar", Name: "Tamnagar, Butwal", Area: "Tamnagar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "nayagaon", Name: "Nayagaon, Butwal", Area: "Nayagaon", City: CityButwal, Province: ProvinceLumbini},
	{ID: "manigram", Name: "Manigram, Butwal", Area: "Manigram", City: CityButwal, Province: ProvinceLumbini},
	{ID: "drivertole", Name: "Drivertole, Butwal", Area: "Drivertole", City: CityButwal, Province: ProvinceLumbini},
	{ID: "yogikuti", Name: "Yogikuti, Butwal", Area: "Yogikuti", City: CityButwal, Province: ProvinceLumbini},
	{ID: "belbas", Name: "Belbas, Butwal", Area: "Belbas", City: CityButwal, Province: ProvinceLumbini},
	{ID: "sukhanagar", Name: "Sukhanagar, Butwal", Area: "Sukhanagar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "amarpath", Name: "Amarpath, Butwal", Area: "Amarpath", City: CityButwal, Province: ProvinceLumbini},
	{ID: "haatbazaar", Name: "Haatbazaar, Butwal", Area: "Haatbazaar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "motipur", Name: "Motipur, Butwal", Area: "Motipur", City: CityButwal, Province: ProvinceLumbini},
	{ID: "semlar", Name: "Semlar, Butwal", Area: "Semlar", City: CityButwal, Province: ProvinceLumbini},
	{ID: "other-butwal", Name: "Other, Butwal", Area: "Other", City: CityButwal, Province: ProvinceLumbini},
}

// GetSupportedLocations returns all supported locations.
func GetSupportedLocations() []LocationItem {
	return ButwalLocations
}

// NormalizeLocation normalizes user input against canonical locations or formats into Title Case.
// For example:
// "devinagar" -> "Devinagar, Butwal"
// "devinagar, butwal" -> "Devinagar, Butwal"
// "lalitpur, kathmandu" -> "Lalitpur, Kathmandu"
func NormalizeLocation(input string) string {
	clean := strings.TrimSpace(input)
	if clean == "" {
		return ""
	}

	lower := strings.ToLower(clean)
	// Check against ButwalLocations by full name or area name
	for _, loc := range ButwalLocations {
		if lower == strings.ToLower(loc.Name) || lower == strings.ToLower(loc.Area) {
			return loc.Name
		}
	}

	// If the user typed "area, butwal" where area matches one of our areas:
	for _, loc := range ButwalLocations {
		if lower == strings.ToLower(loc.Area)+", butwal" {
			return loc.Name
		}
	}

	// Otherwise, apply Title Case formatting
	return utils.FormatLocation(clean)
}
